apps/ex_wire/lib/ex_wire/sync.ex

```diff
@@ -97,47 +130,165 @@ defmodule ExWire.Sync do
+  have new blocks to fetch, so we ask the warp queue for more blocks to
+  request. We may already, however, we waiting on blocks, in which case we
+  do nothing.
+  """
+  @spec handle_snapshot_manifest(SnapshotManifest.t(), Peer.t(), state()) :: state()
+  def handle_snapshot_manifest(%SnapshotManifest{manifest: nil}, _peer, state) do
+    :ok = Logger.debug("Received nil snapshot manifest")
 
-    case packet_res do
-      :ok ->
-        Map.put(state, :last_requested_block, next_block_to_request + @blocks_per_request)
+    state
+  end
 
-      :unsent ->
-        :ok =
-          Logger.debug(fn ->
-            "[Sync] No connected peers to sync, trying again in #{@retry_delay / 1000} second(s)"
-          end)
+  def handle_snapshot_manifest(
+        %SnapshotManifest{manifest: manifest},
+        _peer,
+        state = %{
+          warp_queue: warp_queue,
+          block_tree: block_tree,
+          trie: trie
+        }
+      ) do
+    {next_warp_queue, next_block_tree, next_trie} =
+      warp_queue
+      |> WarpQueue.new_manifest(manifest)
+      |> dispatch_new_warp_queue_requests()
+      |> WarpQueue.process_completed_blocks(block_tree, trie)
+
+    %{
+      state
+      | warp_queue: next_warp_queue,
+        block_tree: next_block_tree,
+        trie: next_trie
+    }
+  end
+
+  @spec dispatch_new_warp_queue_requests(WarpQueue.t(), integer()) :: WarpQueue.t()
+  defp dispatch_new_warp_queue_requests(warp_queue, request_limit \\ @request_limit) do
+    {new_warp_queue, block_hashes_to_request} =
+      WarpQueue.get_block_hashes_to_request(warp_queue, request_limit)
 
-        request_next_block(@retry_delay)
+    Exth.inspect(block_hashes_to_request, "new block hash requestes")
 
-        state
+    for block_hash <- block_hashes_to_request do
+      request_block_chunk(block_hash, true)
     end
+
+    new_warp_queue
+  end
+
+  @doc """
+  When we receive a SnapshotData, let's try to add the received block to the
+  warp queue. We may decide to request new blocks at this time.
+  """
+  @spec handle_snapshot_data(SnapshotData.t(), Peer.t(), state()) :: state()
+  def handle_snapshot_data(%SnapshotData{chunk: nil}, _peer, state) do
+    :ok = Logger.debug("Received empty SnapshotData message.")
+
+    state
+  end
+
+  def handle_snapshot_data(
+        %SnapshotData{chunk: block_chunk = %BlockChunk{}},
+        _peer,
+        state = %{warp_queue: warp_queue}
+      ) do
+    new_warp_queue =
+      warp_queue
+      |> WarpQueue.new_block_chunk(block_chunk)
+      |> dispatch_new_warp_queue_requests(0)
+
+    %{
+      state
+      | warp_queue: new_warp_queue
+    }
+  end
+
+  def handle_snapshot_data(
+        %SnapshotData{chunk: state_chunk = %StateChunk{}},
+        _peer,
+        state = %{warp_queue: _warp_queue}
+      ) do
+    Exth.inspect(state_chunk, "Got state_chunk, not doing what I should")
+
+    # WarpQueue.new_block_chunk(warp_queue, block_chunk)
```

Comment:
Should this comment be removed? Or uncommented?

Reply:
Yeah, this is straight-WIP